src/Lykke.Service.Qtum.Api.Services/HistoryService.cs

```diff
@@ -0,0 +1,188 @@
+﻿using Common.Log;
+using Lykke.Common.Log;
+using Lykke.Service.Qtum.Api.AzureRepositories.Entities.Addresses;
+using Lykke.Service.Qtum.Api.Core.Domain.Addresses;
+using Lykke.Service.Qtum.Api.Core.Domain.InsightApi.AddrTxs;
+using Lykke.Service.Qtum.Api.Core.Helpers;
+using Lykke.Service.Qtum.Api.Core.Repositories.Addresses;
+using Lykke.Service.Qtum.Api.Core.Services;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Lykke.Service.Qtum.Api.Services
+{
+    public class HistoryService<TAddressHistory, TAddressObservation> : IHistoryService<TAddressHistory, TAddressObservation>
+        where TAddressHistory : IAddressHistoryEntry
+        where TAddressObservation : IAddressObservation
+    {
+        private readonly IAddressHistoryEntryRepository<TAddressHistory> _addressHistoryEntryRepository;
+        private readonly IAddressObservationRepository<TAddressObservation> _addressObservationRepository;
+        private readonly IBlockchainService _blockchainService;
+        private readonly IAssetService _assetService;
+        private readonly ILog _log;
+
+        public HistoryService(ILogFactory logFactory, IAddressHistoryEntryRepository<TAddressHistory> addressHistoryEntryRepository, IAddressObservationRepository<TAddressObservation> addressObservationRepository, IBlockchainService blockchainService, IAssetService assetService)
+        {
+            _addressHistoryEntryRepository = addressHistoryEntryRepository;
+            _addressObservationRepository = addressObservationRepository;
+            _blockchainService = blockchainService;
+            _assetService = assetService;
+            _log = logFactory.CreateLog(this);
+        }
+
+        public async Task<bool> AddAddressObservationAsync(TAddressObservation addressObservation)
+        {
+            return await _addressObservationRepository.CreateIfNotExistsAsync(addressObservation);
+        }
+
+        public async Task<(string continuation, IEnumerable<TAddressHistory> items)> GetAddressHistoryAsync(int take, string partitionKey, string address, string afterHash = null, string continuation = null)
+        {
+            if (address != null && partitionKey != null)
+            {
+                if (afterHash != null)
+                {
+                    var afterRecord = await _addressHistoryEntryRepository.GetAsync(afterHash, partitionKey);
+                    var afterBlockCount = afterRecord.BlockCount;
+
+                    return await _addressHistoryEntryRepository.GetByAddressAsync(take, partitionKey, address, afterBlockCount, continuation);
+                }
+                else
+                {
+                    return await _addressHistoryEntryRepository.GetByAddressAsync(take, partitionKey, address, continuation: continuation);
+                }
+            }
+            else
+            {
+                throw new ArgumentException();
```

Comment:
throw different exceptions for address==null and partitionKey==null. It must be ArgumentNullException

Reply:
fixed